cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp

```diff
@@ -0,0 +1,280 @@
+        res_max = reduce(sg, x, maximum<Float>());
+        res_index = reduce(sg, res_max == x ? x_index : int_max, minimum<Float>());
+
+        for (std::uint32_t group_index = sg_size; group_index < sg_count; group_index += sg_size) {
+            x = sg_cache[group_index + sg_local_id].value;
+            x_index = sg_cache[group_index + sg_local_id].index;
+
+            const Float inner_max = reduce(sg, x, maximum<Float>());
+            if (inner_max > res_max) {
+                res_max = inner_max;
+                res_index = reduce(sg, res_max == x ? x_index : int_max, minimum<Float>());
+            }
+        }
+
+        if (sg_local_id == 0) {
+            result.value = res_max;
+            result.index = res_index;
+        }
+    }
+
+    item.barrier(sycl::access::fence_space::local_space);
+}
+
+template <typename Float>
+sycl::event solve_smo(sycl::queue& queue,
+                      const pr::ndview<Float, 1>& kernel_values,
+                      const pr::ndview<std::uint32_t, 1>& ws_indices,
+                      const pr::ndarray<Float, 1>& labels,
+                      const std::int64_t row_count,
+                      const std::int64_t ws_count,
+                      const std::int64_t max_inner_iter,
+                      const Float C,
+                      const Float eps,
+                      const Float tau,
+                      pr::ndview<Float, 1>& alpha,
+                      pr::ndview<Float, 1>& delta_alpha,
+                      pr::ndview<Float, 1>& f,
+                      pr::ndview<Float, 1>& f_diff,
+                      pr::ndview<std::uint32_t, 1>& inner_iter_count,
+                      const dal::backend::event_vector& deps = {}) {
+    ONEDAL_ASSERT(row_count > 0);
+    ONEDAL_ASSERT(row_count <= dal::detail::limits<std::uint32_t>::max());
+    ONEDAL_ASSERT(row_count == labels.get_dimension(0));
+    ONEDAL_ASSERT(max_inner_iter > 0);
+    ONEDAL_ASSERT(max_inner_iter <= dal::detail::limits<std::uint32_t>::max());
+    ONEDAL_ASSERT(ws_count > 0);
+    ONEDAL_ASSERT(ws_count <= dal::detail::limits<std::uint32_t>::max());
+    ONEDAL_ASSERT(ws_indices.get_dimension(0) == ws_count);
+    ONEDAL_ASSERT(delta_alpha.get_dimension(0) == ws_count);
+    ONEDAL_ASSERT(labels.get_dimension(0) == f.get_dimension(0));
+    ONEDAL_ASSERT(alpha.get_dimension(0) == f.get_dimension(0));
+
+    const Float fp_min = dal::detail::limits<Float>::min();
+
+    const Float* labels_ptr = labels.get_data();
+    const Float* kernel_values_ptr = kernel_values.get_data();
+    const std::uint32_t* ws_indices_ptr = ws_indices.get_data();
+    Float* alpha_ptr = alpha.get_mutable_data();
+    Float* delta_alpha_ptr = delta_alpha.get_mutable_data();
+    Float* f_ptr = f.get_mutable_data();
+    Float* f_diff_ptr = f_diff.get_mutable_data();
+    std::uint32_t* inner_iter_count_ptr = inner_iter_count.get_mutable_data();
+
+    constexpr std::uint32_t max_sg_size = 64;
```

Comment:
do you want have as max{wg_size / sg_size, sg_size } ?

Reply:
At submit we don't have information about subgroups, so we can't use this information